### PR TITLE
[TDB-20] added the mtr test from @BohuTang to verify the patch

### PR DESCRIPTION
**Pull request by Jun-Yuan**
[TDB-20] added the MTR test from @BohuTANG to the mysql-test to verify the ft layer changes.

### CODE REVIEW

**Comment by Jun-Yuan** [comment]
Oh god. Please ignore this. This is a mis-operation. I am still waiting for the PerconaFT PR to be approved.

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 1]
```diff
@@ -0,0 +1,266 @@
+# conn1 mocks hot backup
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:1]: Consider putting this in tokudb.backup suite? Although I am not 100% certain, as this is "XB fake backup"

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+--exec cp -rf $MYSQLD_DATADIR/_test_t1* $MYSQL_TMP_DIR/backupdir/
+# fake drop tables during backup
+DROP TABLE test.t22;
+--exec cp -rf $MYSQLD_DATADIR/_test_t2* $MYSQL_TMP_DIR/backupdir/
+DROP TABLE test.t111;
+
+# fake xtrabackup 6) release tokudb checkpoint lock
+SET GLOBAL TOKUDB_CHECKPOINT_LOCK=OFF;
+
+--echo shutdown server
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:70]: Replace by source include/shutdown_mysqld.inc

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+--source include/wait_until_disconnected.inc
+
+# cleanup tokudb files
+--exec rm -rf $MYSQLD_DATADIR/tokudb.*
+--exec rm -rf $MYSQLD_DATADIR/log000*
+
+# fake recovery: copy tokudb.* and redo-log, data files from backup dir
+--exec cp -rf $MYSQL_TMP_DIR/backupdir/* $MYSQLD_DATADIR/
+
+--enable_reconnect
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:82]: Replace by source include/start_mysqld.inc

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+  --dec $i
+}
+--enable_query_log
+
+--disable_query_log
+call mtr.add_suppression("Attempting backtrace");
+--enable_query_log
+
+# make mtr expect the "crash"
+--let $_server_id= `SELECT @@server_id`
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:115]: If this snippet stays in 5.6, just remove and use mysqld.1.expect below, this is not a generic snippet for arbitrary replication testcase server

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  --eval INSERT INTO test.t2 VALUES($i, $i)
+  --dec $i
+}
+--enable_query_log
+
+--disable_query_log
+call mtr.add_suppression("Attempting backtrace");
+--enable_query_log
+
+# make mtr expect the "crash"
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:114]: In the 5.7 version use include/kill_and_restart_mysqld.inc. Consider backporting that include file to 5.6

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+--exec echo "wait" > $_expect_file_name
+--let $_mysqld_datadir= `SELECT @@datadir`
+
+# crashed, sending SIGSEGV
+--let $pid_file_name = `select @@pid_file`
+--exec kill -9 `cat $pid_file_name`;
+--echo 1
+--source include/wait_until_disconnected.inc
+
+--enable_reconnect
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:126]: If this snippet stays in 5.6, replace by start_mysqld.inc

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+DROP TABLE test.t2;
+disconnect conn1;
+
+--echo =====3) create empty table and shutdown with checkpoint lock=====
+connect (conn1, localhost, root,,);
+connection conn1;
+CREATE TABLE test.t1 (c1 INT, c2 INT, PRIMARY KEY(c1), CLUSTERING KEY(c2)) ENGINE=TOKUDB;
+CREATE TABLE test.t2 (c1 INT) ENGINE=TOKUDB;
+
+--echo shutdown server
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:140]: restart_mysqld.inc

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+--enable_reconnect
+--exec echo "restart: --innodb_buffer_pool_size=100M --tokudb_cache_size=100M --tokudb_checkpoint_lock=ON" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+SELECT * FROM test.t1;
+SELECT * FROM test.t2;
+
+CREATE TABLE test.t3 (c1 INT) ENGINE=TOKUDB;
+INSERT INTO test.t3 VALUES(1);
+
+--echo shutdown server
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:154]: Likewise

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+ALTER TABLE test.t1 ADD COLUMN `c3` INT;
+INSERT INTO test.t1 VALUES(2,2,3);
+
+connection conn1;
+# copy tokudb files after table altered
+--exec cp -rf $MYSQLD_DATADIR/_test_t1* $MYSQL_TMP_DIR/backupdir/
+
+SET GLOBAL TOKUDB_CHECKPOINT_LOCK=OFF;
+
+--echo shutdown server
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:196]: Likewise

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+--enable_reconnect
+--exec echo "restart: --innodb_buffer_pool_size=100M --tokudb_cache_size=100M" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+
+OPTIMIZE TABLE test.t1;
+SELECT * FROM test.t1;
+
+DROP TABLE test.t1;
+disconnect conn1;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:217]: conn1 / conn2 could not have survived restart (if they did, courtesy of enable_reconnect, then it's still redundant)

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+disconnect conn2;
+--exec rm -rf $MYSQL_TMP_DIR/backupdir/
+
+--echo =====5) kill insert/alter/add index when checkpoint lock ON=====
+connect (conn1, localhost, root,,);
+connection conn1;
+--echo shutdown server
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:224]: MTR include files again

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+connection conn2;
+CREATE TABLE test.t3 (c1 INT PRIMARY KEY, c2 INT) ENGINE=TOKUDB;
+ALTER TABLE test.t1 ADD COLUMN `c3` INT;
+INSERT INTO test.t1 VALUES(2,2,3);
+
+connect (conn3, localhost, root,,);
+connection conn3;
+ALTER TABLE test.t3 ADD INDEX idx_c1c2(c1, c2);
+
+# crashed, sending SIGSEGV
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:247]: Likewise

File: mysql-test/suite/tokudb/t/in_backup.test [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+--enable_reconnect
+# start mysqld
+--exec echo "restart" > $_expect_file_name
+--source include/wait_until_connected_again.inc
+SHOW CREATE TABLE test.t1;
+SHOW CREATE TABLE test.t3;
+DROP TABLE test.t1;
+DROP TABLE test.t3;
+disconnect conn1;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/tokudb/t/in_backup.test:264]: Should be no need to disconnect after restart

**Review by laurynas-biveinis — COMMENTED** [review]
I have reviewed the testcase nevertheless - feel free to ignore/address to the extent it will apply to the final PR

**Pull request closed without merging**